src/publisher/index.ts

```diff
@@ -0,0 +1,164 @@
+/*!
+ * Copyright 2019 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as extend from 'extend';
+import {CallOptions} from 'google-gax';
+
+import {MessageBatch, BatchPublishOptions} from './message-batch';
+import {Queue, OrderedQueue} from './message-queues';
+import {Topic} from '../topic';
+import {RequestCallback} from '../pubsub';
+import {google} from '../../proto/pubsub';
+
+export type PubsubMessage = google.pubsub.v1.IPubsubMessage;
+
+export interface Attributes {
+  [key: string]: string;
+}
+
+export type PublishCallback = RequestCallback<string>;
+
+export interface PublishOptions {
+  batching?: BatchPublishOptions;
+  gaxOpts?: CallOptions;
+  messageOrdering?: boolean;
+}
+
+/**
+ * A Publisher object allows you to publish messages to a specific topic.
+ *
+ * @private
+ * @class
+ *
+ * @see [Topics: publish API Documentation]{@link https://cloud.google.com/pubsub/docs/reference/rest/v1/projects.topics/publish}
+ *
+ * @param {Topic} topic The topic associated with this publisher.
+ * @param {PublishOptions} [options] Configuration object.
+ */
+export class Publisher {
+  // tslint:disable-next-line variable-name
+  Promise?: PromiseConstructor;
+  topic: Topic;
+  settings!: PublishOptions;
+  queue: Queue;
+  orderedQueues: Map<string, OrderedQueue>;
+  constructor(topic: Topic, options?: PublishOptions) {
+    if (topic.Promise) {
+      this.Promise = topic.Promise;
+    }
+
+    this.setOptions(options);
+    this.topic = topic;
+    this.queue = new Queue(this);
+    this.orderedQueues = new Map();
+  }
+  /**
+   * Publish the provided message.
+   *
+   * @private
+   *
+   * @throws {TypeError} If data is not a Buffer object.
+   * @throws {TypeError} If any value in `attributes` object is not a string.
+   *
+   * @param {buffer} data The message data. This must come in the form of a
+   *     Buffer object.
+   * @param {MessageOptions} [options] Options for this message.
+   * @param {PublishCallback} [callback] Callback function.
+   */
+  publish(message: PubsubMessage, callback: PublishCallback): void {
```

Comment:
The parameters don't seem to match the description in the comment.

Reply:
Ah! Nice find